### PR TITLE
feat: add property tests for matrix decompositions

### PR DESCRIPTION
**Pull request by polvalente**
Adds some property tests as discussed in #410

### CODE REVIEW

File: nx/test/nx/lin_alg_test.exs [hunk 1]
```diff
@@ -186,15 +200,93 @@ defmodule Nx.LinAlgTest do
...
+        m = u |> Nx.shape() |> elem(1)
+        n = vt |> Nx.shape() |> elem(0)
+
+        assert u
+               |> Nx.dot(diag(d, m, n))
+               |> Nx.dot(vt)
+               |> Nx.subtract(tall)
+               |> Nx.all_close?(1.0e-5)
+
+        # TODO: SVD does not work for wide matrices and
```
Comment (polvalente) [inline nx/test/nx/lin_alg_test.exs:230]: I think this should be fixed in a subsequent PR instead of this one

File: nx/test/nx/defn/grad_test.exs [hunk 2]
```diff
@@ -1716,9 +1716,14 @@ defmodule Nx.Defn.GradTest do
 
   describe "slice" do
     defn grad_mean_slice(t), do: grad(t, &Nx.mean(Nx.slice(&1, [0, 1], [1, 2], strides: [1, 2])))
-    defn grad_mean_dynamic_slice(t), do: grad(t, &Nx.mean(Nx.slice(&1, [Nx.tensor(0), Nx.tensor(1)], [1, 2], strides: [1, 2])))
+
```
Comment (polvalente) [inline nx/test/nx/defn/grad_test.exs:1719]: Let me know if I should revert these stray formatting changes

**Review by seanmor5 — APPROVED** [review]

File: nx/lib/nx/shape.ex [hunk 3]
```diff
@@ -1370,14 +1370,18 @@ defmodule Nx.Shape do
   def solve({n, n}, b_shape) do
     raise(
       ArgumentError,
-      "`b` tensor has incompatible dimensions, expected #{inspect({n, n})} or {#{n}}, got: #{inspect(b_shape)}"
+      "`b` tensor has incompatible dimensions, expected #{inspect({n, n})} or {#{n}}, got: #{
+        inspect(b_shape)
+      }"
```
Comment (josevalim) [inline nx/lib/nx/shape.ex:1375]:
```suggestion
      "`b` tensor has incompatible dimensions, expected #{inspect({n, n})} or {#{n}}, got: " <>
        inspect(b_shape)
```

File: nx/lib/nx/shape.ex [hunk 4]
```diff
@@ -1370,14 +1370,18 @@ defmodule Nx.Shape do
...
     )
   end
 
   def solve(a_shape, _b_shape) do
     raise(
       ArgumentError,
-      "`a` tensor has incompatible dimensions, expected a 2-D tensor with as many rows as columns, got: #{inspect(a_shape)}"
+      "`a` tensor has incompatible dimensions, expected a 2-D tensor with as many rows as columns, got: #{
+        inspect(a_shape)
+      }"
```
Comment (josevalim) [inline nx/lib/nx/shape.ex:1384]:
```suggestion
      "`a` tensor has incompatible dimensions, expected a 2-D tensor with as many rows as columns, got: " <>
        inspect(a_shape)
```

File: nx/lib/nx/tensor.ex [hunk 5]
```diff
@@ -115,7 +115,9 @@ defmodule Nx.Tensor do
 
       {{^axis, value}, _} ->
         raise ArgumentError,
-              "slicing a tensor on an axis requires an integer, a scalar tensor or a range, got: #{inspect(value)}"
+              "slicing a tensor on an axis requires an integer, a scalar tensor or a range, got: #{
+                inspect(value)
+              }"
```
Comment (josevalim) [inline nx/lib/nx/tensor.ex:120]:
```suggestion
              "slicing a tensor on an axis requires an integer, a scalar tensor or a range, got: " <>
                inspect(value)
```

File: nx/test/nx/lin_alg_test.exs [hunk 6]
```diff
@@ -186,15 +200,93 @@ defmodule Nx.LinAlgTest do
              ])
              |> round(3) == round(v, 3)
     end
+
+    test "reconstruction check" do
```
Comment (josevalim) [inline nx/test/nx/lin_alg_test.exs:204]:
```suggestion
    test "property" do
```

File: nx/test/nx/lin_alg_test.exs [hunk 7]
```diff
@@ -108,6 +108,20 @@ defmodule Nx.LinAlgTest do
                [0.0, 0.0, 0.0]
              ]) == q |> Nx.dot(r) |> round(1)
     end
+
+    test "reconstruction check" do
```
Comment (josevalim) [inline nx/test/nx/lin_alg_test.exs:112]:
```suggestion
    test "property" do
```

**Review by josevalim — APPROVED** [review]
Minor comments and ship it!!!!

**Pull request merged**